### ISSUE TITLE
Hammox slow with stripity_stripe lib

### ISSUE THREAD

**Issue by jotaviobiondo** [msz/hammox#47]
When using `Hammox.expect` returning a struct from [stripity_stripe](https://github.com/code-corps/stripity_stripe), like the [`Stripe.Customer`](https://hexdocs.pm/stripity_stripe/Stripe.Customer.html#t:t/0) struct, it becomes too slow (more or less 5 seconds for each `expect` call).

Seems to be an issue with the typespec checking (after returning the value), because if I use `Hammox.stub` or even use `Mox.expect`, it runs normally.

An example of the call I'am doing:
```elixir
Hammox.expect(StripeMock, :get_customer, fn _ -> 
  %Stripe,Customer{
      id: "",
      object: "customer",
      address: nil,
      balance: 0,
      created: 1_599_741_039,
      currency: "usd",
      default_source: "",
      deleted: nil,
      delinquent: false,
      description: nil,
      discount: nil,
      email: nil,
      invoice_prefix: "",
      invoice_settings: nil,
      livemode: false,
      metadata: %{},
      name: "",
      next_invoice_sequence: 1,
      payment_method: nil,
      phone: "",
      preferred_locales: [],
      shipping: nil,
      sources: %List{data: [], has_more: false, object: "", total_count: 0, url: ""},
      subscriptions: %List{data: [], has_more: false, object: "", total_count: 0, url: ""},
      tax_exempt: nil,
      tax_ids: %List{data: [], has_more: false, object: "", total_count: 0, url: ""},
    }
end)
```

Using 
- Erlang/OTP 23 
- Elixir 1.10.4
- Hammox 0.2.5
- stripity_stripe: 2.9

**Comment by msz** [comment]
Thanks for reporting this!

The Hammox typespec checker is still in its "first draft" performance wise — performance simply hasn't been considered at all yet as it worked quite fine for my personal purposes until now. I'm sure there is some low hanging fruit to pick. Now that I have a slow case, I'll take a stab at investigating in detail!

**Comment by jotaviobiondo** [comment]
Nice! It worked fine for the company I work too, until now.

Congrats on the lib, love the idea of checking the typespec :)

**Comment by randycoulman** [comment]
I've found Hammox to be quite slow on our (private, unfortunately) test suite. I've (off and on) been trying to find a way to run a profiler against one of our slow tests to get a sense of where the issue might be before reporting it here.  It seems to consistently run the same set of tests slower, but I haven't found a common suspect in the slow tests.  It doesn't seem to be a general slow-down across all of our tests from what I can tell.

If I find a way to get some more useful information than this, I'll definitely report back.

But for now, I'll offer to test out any fixes you come up with to see if they improve our situation.

I really love this library and want to make it part of our toolchain. Every time I go back to my hammox branch and rebase, it catches something for us.  But right now it slows our test suite down too much (~150s vs ~30s) for me to be able to recommend that we use it.

**Comment by msz** [comment]
Thanks for the input @randycoulman! That slowdown is really unacceptable. I will prioritise solving this higher.

**Comment by randycoulman** [comment]
Thanks @msz! If there's anything I can do to help, or tests you want me to run, I'd be happy to.

I tried once again to get a profile of a single test run, but it failed to complete after 21 hours and I had to kill it so I could get some other things done.  The test normally takes ~10s to run with Hammox.

**Comment by msz** [comment]
I did some profiling at it looks like the big bottleneck is repeatedly fetching types from disk which should be easy to solve by introducing a cache. Will work on it this weekend.

**Comment by SophisticaSean** [comment]
@msz Thanks for looking into this. I wanted to also pipe up and say I've noticed this as well. With Hammox my 1 test runs in 11ish seconds, just using Mox it runs in 13.1ms. I'm happy to help optimize this anyway I can, just let me know how I can help.

**Comment by SophisticaSean** [comment]
@msz I'm also happy to test any rough drafts you have. I have a test file that is 7 tests long and runs in 55 seconds on Hammox and 0.7 seconds on Mox.

**Comment by msz** [comment]
I've just merged a PR containing a performance boost which you might all find quite satisfactory. Could you try with latest master?

**Comment by randycoulman** [comment]
@msz I just ran the master version on our test suite, and wow!  There is no difference in test run time with and without Hammox now.  Nice job!

**Comment by jotaviobiondo** [comment]
Nice job! I also tested it here on our test suite and couldn't see any difference from Mox!

**Comment by msz** [comment]
Amazing! I consider this fixed by #53 then.
